### PR TITLE
Remove the possibility of using 'localhost' in KubeService ExternalName

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Removes the possibility of using localhost in ExternalNames as endpoints

# Context

I also updated the README for our regression tests to make local development (and testing) easier.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: ci/deploy-to-kind-cluster.sh [hunk 1]
```diff
@@ -32,29 +48,23 @@ kubeadmConfigPatches:
...
+  echo "Finished setting up cluster $CLUSTER_NAME"
 }
-trap "cleanup" EXIT SIGINT
+create_kind_cluster_or_skip
 
-echo ">> Temporary output file ${TEMP_FILE}"
+# 2. Make all the docker images and load them to the kind cluster
+VERSION=$VERSION CLUSTER_NAME=$CLUSTER_NAME make push-kind-images
 
-# grab the image names out of the `make docker` output
```
Comment (sam-heilbron) [inline ci/deploy-to-kind-cluster.sh:49]: Previously, we relied on the output of the `make docker` command to extract the images and load them into the kind cluster. In newer docker versions however, this output does not exist and this command will stop working.

File: projects/ingress/pkg/status/status_syncer.go [hunk 2]
```diff
@@ -96,9 +103,15 @@ func getLbStatus(services v1.KubeServiceList) ([]kubev1.LoadBalancerIngress, err
...
 
-func serviceAddrs(svc *kubev1.Service) []string {
+func serviceAddrs(svc *kubev1.Service, kubeSvcRef *core.ResourceRef) ([]string, error) {
 	if svc.Spec.Type == kubev1.ServiceTypeExternalName {
-		return []string{svc.Spec.ExternalName}
+
+		// Remove the possibility of using localhost in ExternalNames as endpoints
+		svcIp := net.ParseIP(svc.Spec.ExternalName)
+		if svc.Spec.ExternalName == "localhost" || (svcIp != nil && svcIp.IsLoopback()) {
+			return nil, errors.Errorf("Invalid attempt to use localhost name %s, in %q", svc.Spec.ExternalName, kubeSvcRef)
```
Comment (kdorosh) [inline projects/ingress/pkg/status/status_syncer.go:112]: did you mean to do `%q` here? (im reading it as a single character: https://golang.org/pkg/fmt/)

File: projects/ingress/pkg/status/status_syncer_test.go [hunk 3]
```diff
@@ -32,39 +32,44 @@ var _ = Describe("StatusSyncer", func() {
...
 		Expect(err).NotTo(HaveOccurred())
 
-		kube, err := kubernetes.NewForConfig(cfg)
+		// Initialize the kube Clientset
+		kubeClientset, err = kubernetes.NewForConfig(cfg)
 		Expect(err).NotTo(HaveOccurred())
-		_, err = kube.CoreV1().Namespaces().Create(ctx, &kubev1.Namespace{
+
+		// Create test namespace
+		namespace = helpers.RandString(8)
```
Comment (kdorosh) [inline projects/ingress/pkg/status/status_syncer_test.go:55]: where was the namespace being set before? why this change?

File: projects/ingress/pkg/status/status_syncer_test.go [hunk 4]
```diff
@@ -191,4 +196,133 @@ var _ = Describe("StatusSyncer", func() {
...
+		statusEmitter := v1.NewStatusEmitter(kubeServiceClient, ingressClient)
+		statusSync := status.NewSyncer(ingressClient)
+		statusEventLoop := v1.NewStatusEventLoop(statusEmitter, statusSync)
+		statusEventLoopErrs, err := statusEventLoop.Run([]string{namespace}, clients.WatchOpts{Ctx: context.TODO()})
+		Expect(err).NotTo(HaveOccurred())
+		go func() {
+			defer GinkgoRecover()
+			err := <-statusEventLoopErrs
+			// Expect an error to have occurred during the statusEventLoop
+			Expect(err).To(HaveOccurred())
```
Comment (kdorosh) [inline projects/ingress/pkg/status/status_syncer_test.go:217]: perhaps would have been more idiomatic here to use `Expect(statusEventLoopErrs).To(HaveRecevied()))`

File: projects/ingress/pkg/status/status_syncer_test.go [hunk 5]
```diff
@@ -191,4 +196,133 @@ var _ = Describe("StatusSyncer", func() {
...
+						Image: "nginx:latest",
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		time.Sleep(time.Second) // give the kube service time to update lb endpoints
+		kubeSvc, err = kubeSvcClient.Get(ctx, kubeSvc.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
```
Comment (kdorosh) [inline projects/ingress/pkg/status/status_syncer_test.go:311]: eventually here instead?

File: projects/ingress/pkg/status/status_syncer_test.go (hunk 3 above)
Comment (sam-heilbron) [reply projects/ingress/pkg/status/status_syncer_test.go:55]: I had just moved this down a few lines from https://github.com/solo-io/gloo/pull/4970/files#diff-6f555b822e955aa09004f740235b0a9102d611a99c71c7bdc8ff9f1bd3c2c969L41

File: projects/ingress/pkg/status/status_syncer.go (hunk 2 above)
Comment (sam-heilbron) [reply projects/ingress/pkg/status/status_syncer.go]: good catch!

File: projects/ingress/pkg/status/status_syncer_test.go (hunk 4 above)
Comment (sam-heilbron) [reply projects/ingress/pkg/status/status_syncer_test.go]: I tried using Receive but ran into some challenges. Also, when I used `Eventually...Receive` the behavior was inconsistent since we start the goroutine right away, and then perform some other kube reads/writes before the error even populates.

File: projects/ingress/pkg/status/status_syncer_test.go (hunk 3 above)
Comment (sam-heilbron) [reply projects/ingress/pkg/status/status_syncer_test.go:55]: Perhaps an unnecessary move. I was surprised that we were writing to a unique namespace each time, and felt that I should make it clearer in the test

**Review by kdorosh — APPROVED** [review]
nice!
